File: R/scpi/tests/testthat/test-scdataMulti.R
library(scpi)

###############################################################################
###############################################################################
## Auxiliary functions for testing

test.data <- function(df,
                      features = NULL,
                      cov.adj = NULL,
                      cointegrated.data = FALSE,
                      post.est = NULL,
                      units.est = NULL,
                      donors.est = NULL,
                      anticipation = 0,
                      effect = "unit-time",
                      constant = FALSE,
                      verbose = FALSE,
                      sparse.matrices = FALSE) {
  data <- scpi_germany
  
  # Create a second placebo treated unit
  data$treatment <- 0
  data[(data$country == "West Germany" & data$year >= 1991), "treatment"] <- 1
  data[(data$country == "Italy" & data$year >= 1992), "treatment"] <- 1
  
  df <- scdataMulti(data, id.var = "country", outcome.var = "gdp",
                    treatment.var = "treatment", time.var = "year", constant = constant,
                    cointegrated.data = cointegrated.data, features = features,
                    cov.adj = cov.adj, verbose = verbose, post.est = post.est,
                    units.est = units.est, donors.est = donors.est, anticipation = anticipation,
                    effect = effect, sparse.matrices = sparse.matrices)
  
  return(df)
}

###############################################################################
###############################################################################
## Auxiliary functions for testing

test_that("no error is returned", {
  expect_no_error(test.data(constant = TRUE, sparse.matrices=FALSE, effect="unit-time",
                  cointegrated.data = TRUE, features = list(c("gdp", "trade")),
                  cov.adj = list(c("constant", "trend"))))
  expect_no_error(test.data(constant = TRUE, sparse.matrices=FALSE, effect="time",
                  cointegrated.data = TRUE, features = list(c("gdp", "trade")),
                  cov.adj = list(c("constant", "trend"))))
  expect_no_error(test.data(constant = TRUE, sparse.matrices=FALSE, effect="unit",
                  cointegrated.data = TRUE, features = list(c("gdp", "trade")),
                  cov.adj = list(c("constant", "trend"))))
  expect_no_error(test.data(constant = TRUE, sparse.matrices=FALSE, effect="unit", post=5,
                            cointegrated.data = TRUE, features = list(c("gdp", "trade")),
                            cov.adj = list(c("constant", "trend"))))
})

File: Python/scpi_pkg/tests/test_scdataMulti.py
from pathlib import Path

import numpy as np
import pandas as pd

from scpi_pkg.scdataMulti import scdataMulti
from scpi_pkg.scpi import scpi


def test_scdata_multi_time_effect_uses_treatment_date_label():
    data = pd.read_csv(Path(__file__).resolve().parents[2] / "scpi_germany.csv")
    data["status"] = 0
    data.loc[(data["country"] == "West Germany") & (data["year"] >= 1991), "status"] = 1
    data.loc[(data["country"] == "Italy") & (data["year"] >= 1992), "status"] = 1

    covs_adj = {
        "Italy": ["constant", "trend"],
        "West Germany": [["constant", "trend"], ["constant", "trend"]],
    }

    result = scdataMulti(
        df=data,
        id_var="country",
        treatment_var="status",
        outcome_var="gdp",
        time_var="year",
        features={
            "Italy": ["gdp", "trade"],
            "West Germany": ["gdp", "infrate"],
        },
        constant={"Italy": True, "West Germany": False},
        cointegrated_data=True,
        cov_adj=covs_adj,
        effect="time",
        verbose=False,
    )

    assert result.effect == "time"
    assert result.A.shape == (125, 1)
    assert result.Y_post.shape == (25, 1)


def test_scpi_unit_effect_feature_designs_stay_conformable():
    data = pd.read_csv(Path(__file__).resolve().parents[2] / "scpi_germany.csv")
    data["status"] = 0
    data.loc[(data["country"] == "West Germany") & (data["year"] >= 1991), "status"] = 1
    data.loc[(data["country"] == "Italy") & (data["year"] >= 1992), "status"] = 1

    result = scdataMulti(
        df=data,
        id_var="country",
        treatment_var="status",
        outcome_var="gdp",
        time_var="year",
        features={"features": ["gdp", "trade"]},
        cov_adj={"cov_adj": [["constant"], ["trend"]]},
        constant=False,
        cointegrated_data=True,
        effect="unit",
        verbose=False,
    )

    np.random.seed(8894)
    intervals = scpi(result, sims=10, cores=1, verbose=False)

    assert len(intervals.CI_in_sample) == len(result.P)
